File: firmware/src/filter.cpp
#include "filter.hpp"
#include <cmath>

ReadingFilter::ReadingFilter(std::chrono::milliseconds interval, 
                             double delta_temp, 
                             double delta_humidity, 
                             double delta_pressure)
    : interval_(interval),
      delta_temp_(delta_temp),
      delta_humidity_(delta_humidity),
      delta_pressure_(delta_pressure),
      last_reading_{0.0, 0.0, 0.0, 0},
      last_published_time_(0),
      has_last_reading_(false) {}

bool ReadingFilter::should_publish(const SensorReading& reading, std::chrono::milliseconds current_time) {
    if (!has_last_reading_) {
        last_reading_ = reading;
        last_published_time_ = current_time;
        has_last_reading_ = true;
        return true;
    }

    auto elapsed = current_time - last_published_time_;
    if (elapsed >= interval_) {
        last_reading_ = reading;
        last_published_time_ = current_time;
        return true;
    }

    double temp_diff = std::fabs(reading.temperature - last_reading_.temperature);
    double hum_diff = std::fabs(reading.humidity - last_reading_.humidity);
    double press_diff = std::fabs(reading.pressure - last_reading_.pressure);

    if (temp_diff >= delta_temp_ || hum_diff >= delta_humidity_ || press_diff >= delta_pressure_) {
        last_reading_ = reading;
        last_published_time_ = current_time;
        return true;
    }

    return false;
}

File: environment/filter.py
from __future__ import annotations

import time

from environment.sensor import SensorReading


class ReadingFilter:
    """Filters sensor readings based on time intervals and value deltas."""

    def __init__(
        self,
        interval_seconds: float = 300.0,
        delta_temp: float = 0.5,
        delta_humidity: float = 2.0,
        delta_pressure: float = 1.0,
    ) -> None:
        self.interval_seconds = interval_seconds
        self.delta_temp = delta_temp
        self.delta_humidity = delta_humidity
        self.delta_pressure = delta_pressure
        self.last_reading: SensorReading | None = None
        self.last_published_time: float = 0.0

    def should_publish(self, reading: SensorReading, current_time: float | None = None) -> bool:
        now = current_time if current_time is not None else time.time()

        if self.last_reading is None:
            self.last_reading = reading
            self.last_published_time = now
            return True

        elapsed = now - self.last_published_time
        if elapsed >= self.interval_seconds:
            self.last_reading = reading
            self.last_published_time = now
            return True

        temp_diff = abs(reading.temperature - self.last_reading.temperature)
        hum_diff = abs(reading.humidity - self.last_reading.humidity)
        press_diff = abs(reading.pressure - self.last_reading.pressure)

        if (
            temp_diff >= self.delta_temp
            or hum_diff >= self.delta_humidity
            or press_diff >= self.delta_pressure
        ):
            self.last_reading = reading
            self.last_published_time = now
            return True

        return False
